updater_test.go

```diff
@@ -470,6 +490,253 @@ func TestUpdaterNotNeeded(t *testing.T) {
 	assert.Equal(t, "deadbeef", upr.config.GetInstallID())
 }
 
+func TestUpdaterCheckAndUpdate(t *testing.T) {
+	testServer := testServerForUpdateFile(t, testZipPath)
+	defer testServer.Close()
+
+	testUpdate := newTestUpdate(testServer.URL, false)
+	upr, err := newTestUpdaterWithServer(t, testServer, testUpdate, &testConfig{})
+	assert.NoError(t, err)
+	defer func() {
+		err = upr.CleanupPreviousUpdates()
+		assert.NoError(t, err)
+	}()
+	ctx := newTestContext(newDefaultTestUpdateOptions(), upr.config, &UpdatePromptResponse{Action: UpdateActionSnooze, AutoUpdate: true})
+
+	// 1.No update from the server
+	// Need update = false
+	// FindDownloadedAsset = false
+	// return updateAvailable = false, updateCached = false
+	updateAvailable, updateCached, err := upr.CheckAndDownload(ctx)
+	assert.NoError(t, err)
+	assert.False(t, updateAvailable)
+	assert.False(t, updateCached)
+	assert.False(t, ctx.successReported)
+	assert.Equal(t, "deadbeef", upr.config.GetInstallID())
+
+	testUpdate.NeedUpdate = true
+
+	// 2. Download asset from URL
+	// Need update = true
+	updateAvailable, updateCached, err = upr.CheckAndDownload(ctx)
+	assert.NoError(t, err)
+	assert.True(t, updateAvailable)
+	assert.True(t, updateCached)
+	assert.False(t, ctx.successReported)
+	assert.Equal(t, "deadbeef", upr.config.GetInstallID())
+
+	// 3.Find existing downloaded asset
+	// Need update = true
+	// FindDownloadedAsset = true
+	// return updateAvailable = true, updateCached = true
```

Comment:
we should check the the asset isn't re-downloaded in the case

Reply:
Good catch, thanks!